<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Founders · XSim.jl</title><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.039/juliamono-regular.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.15.3/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.13.11/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../../"><img src="../../assets/logo.png" alt="XSim.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../../">XSim.jl</a></span></div><form class="docs-search" action="../../search/"><input class="docs-search-query" id="documenter-search-query" name="q" type="text" placeholder="Search docs"/></form><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><a class="tocitem" href="../../demo/">Getting Started</a></li><li><span class="tocitem">Basic Usages</span><ul><li><a class="tocitem" href="../build_genome/">Build Genome</a></li><li><a class="tocitem" href="../build_phenome/">Build Phenome</a></li><li class="is-active"><a class="tocitem" href>Founders</a><ul class="internal"><li><a class="tocitem" href="#Data"><span>Data</span></a></li><li><a class="tocitem" href="#By-Assigning-a-Population-Size"><span>By Assigning a Population Size</span></a></li><li><a class="tocitem" href="#By-a-genoetype-or-haplotypes-file"><span>By a genoetype or haplotypes file</span></a></li><li><a class="tocitem" href="#Inspect-the-founders"><span>Inspect the founders</span></a></li></ul></li><li><a class="tocitem" href="../mate/">Mate</a></li><li><a class="tocitem" href="../select/">Select</a></li><li><a class="tocitem" href="../breed/">Breed</a></li></ul></li><li><span class="tocitem">Case Studies</span><ul><li><a class="tocitem" href="../../case/crossbreed/">Rotational Cross-Breeding</a></li><li><a class="tocitem" href="../../case/NAM/">NAM</a></li></ul></li><li><span class="tocitem">Genetic Evaluation</span><ul><li><a class="tocitem" href="../../ge/gblup_pblup/">GBLUP and PBLUP Evaluation</a></li><li><a class="tocitem" href="../../ge/multi_trait/">Multi-Trait Genetic Evaluation</a></li></ul></li><li><span class="tocitem">For Developers</span><ul><li><a class="tocitem" href="../../dev/custom_data/">Customized Data</a></li><li><a class="tocitem" href="../../dev/expand_genome/">Expand Genome Information</a></li></ul></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Basic Usages</a></li><li class="is-active"><a href>Founders</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Founders</a></li></ul></nav><div class="docs-right"><a class="docs-edit-link" href="https://github.com/Poissonfish/XSim.jl/blob/master/docs/src/basic/founder.md" title="Edit on GitHub"><span class="docs-icon fab"></span><span class="docs-label is-hidden-touch">Edit on GitHub</span></a><a class="docs-settings-button fas fa-cog" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-sidebar-button fa fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a></div></header><article class="content" id="documenter-page"><h1 id="Founders"><a class="docs-heading-anchor" href="#Founders">Founders</a><a id="Founders-1"></a><a class="docs-heading-anchor-permalink" href="#Founders" title="Permalink"></a></h1><h2 id="Data"><a class="docs-heading-anchor" href="#Data">Data</a><a id="Data-1"></a><a class="docs-heading-anchor-permalink" href="#Data" title="Permalink"></a></h2><p>In this page, Files <code>demo_genotypes.csv</code>, <code>demo_haplotypes.csv</code> will be used in the demonstrated examples. Missing values can be represented by -1 or 9. Both demonstrated files contain 5 individuals and 4 genetic markers.</p><pre><code class="nohighlight"># demo_genotypes.csv
# rows: individuals, columns: markers
# homozygote is coded as 0 and 2, otherwise is coded as 1
2,0,0,1
0,0,1,0
0,1,0,2
1,1,0,2
2,0,2,0

# demo_haplotypes.csv
# rows: individuals, columns: alleles
# Reference allele is coded as 0, otherwise is coded as 1
1,1,0,0,0,0,1,0
0,0,0,0,1,0,0,0
0,0,0,1,0,0,1,1
1,0,1,0,0,0,1,1
1,1,0,0,1,1,0,0</code></pre><h2 id="By-Assigning-a-Population-Size"><a class="docs-heading-anchor" href="#By-Assigning-a-Population-Size">By Assigning a Population Size</a><a id="By-Assigning-a-Population-Size-1"></a><a class="docs-heading-anchor-permalink" href="#By-Assigning-a-Population-Size" title="Permalink"></a></h2><pre><code class="language-julia">founders = Founders(5)</code></pre><pre><code class="nohighlight">[ Info: Cohort (5 individuals)
[ Info:
[ Info: Mean of breeding values:
[ Info: [1.265 1.697]
[ Info:
[ Info: Variance of breeding values:
[ Info: [1.6 1.4]</code></pre><h2 id="By-a-genoetype-or-haplotypes-file"><a class="docs-heading-anchor" href="#By-a-genoetype-or-haplotypes-file">By a genoetype or haplotypes file</a><a id="By-a-genoetype-or-haplotypes-file-1"></a><a class="docs-heading-anchor-permalink" href="#By-a-genoetype-or-haplotypes-file" title="Permalink"></a></h2><pre><code class="language-julia"># haplotypes
founder = Founders(&quot;demo_haplotypes.csv&quot;)</code></pre><pre><code class="nohighlight">[ Info: MAF has been updated based on provided haplotypes/genotypes
[ Info: Cohort (5 individuals)
[ Info: 
[ Info: Mean of breeding values: 
[ Info: [1.391 0.922]
[ Info: 
[ Info: Variance of breeding values: 
[ Info: [2.08 0.747]</code></pre><p>Or if users don&#39;t want to update the MAF by the data, <code>alter_maf=false</code> can achieve this purpose.</p><pre><code class="language-julia"># genotypes
founders = Founders(&quot;demo_genotypes.csv&quot;, alter_maf=false)</code></pre><pre><code class="nohighlight">[ Info: Cohort (5 individuals)
[ Info: 
[ Info: Mean of breeding values: 
[ Info: [1.391 0.922]
[ Info: 
[ Info: Variance of breeding values: 
[ Info: [2.08 0.747]</code></pre><p>Users can subset the population by assigning <code>n</code> and <code>random</code>. If <code>random=true</code>, <code>n</code> individuals are randomly selected from the file. Otherwise, only the first <code>n</code> individuals will be initialized.</p><pre><code class="language-julia">founders = Founders(&quot;demo_genotypes.csv&quot;, n=3, random=true)</code></pre><pre><code class="nohighlight">[ Info: MAF has been updated based on provided haplotypes/genotypes
[ Info: Cohort (3 individuals)
[ Info: 
[ Info: Mean of breeding values: 
[ Info: [2.108 0.404]
[ Info: 
[ Info: Variance of breeding values: 
[ Info: [2.133 0.49]</code></pre><h2 id="Inspect-the-founders"><a class="docs-heading-anchor" href="#Inspect-the-founders">Inspect the founders</a><a id="Inspect-the-founders-1"></a><a class="docs-heading-anchor-permalink" href="#Inspect-the-founders" title="Permalink"></a></h2><h3 id="Genotypes"><a class="docs-heading-anchor" href="#Genotypes">Genotypes</a><a id="Genotypes-1"></a><a class="docs-heading-anchor-permalink" href="#Genotypes" title="Permalink"></a></h3><pre><code class="language-julia">get_genotypes(founders)</code></pre><pre><code class="nohighlight">5×4 LinearAlgebra.Adjoint{Int64,Array{Int64,2}}:
 0  0  1  0
 2  0  2  0
 2  0  0  1
 0  1  0  2
 1  1  0  2</code></pre><h3 id="QTLs"><a class="docs-heading-anchor" href="#QTLs">QTLs</a><a id="QTLs-1"></a><a class="docs-heading-anchor-permalink" href="#QTLs" title="Permalink"></a></h3><pre><code class="language-julia">get_QTLs(founders)</code></pre><pre><code class="nohighlight">5×3 Array{Int64,2}:
 2  2  0
 0  0  2
 0  1  0
 1  0  2
 2  0  1</code></pre><h3 id="Breeding-Values"><a class="docs-heading-anchor" href="#Breeding-Values">Breeding Values</a><a id="Breeding-Values-1"></a><a class="docs-heading-anchor-permalink" href="#Breeding-Values" title="Permalink"></a></h3><p>It&#39;s stored as a n by n_traits matrix</p><pre><code class="language-julia">get_BVs(founders)</code></pre><pre><code class="nohighlight">5×2 LinearAlgebra.Adjoint{Float64,Array{Float64,2}}:
 1.26491   0.0
 3.79473   0.0
 1.26491   1.21268
 0.0       1.69775
 0.632456  1.69775</code></pre><h3 id="Pedigree"><a class="docs-heading-anchor" href="#Pedigree">Pedigree</a><a id="Pedigree-1"></a><a class="docs-heading-anchor-permalink" href="#Pedigree" title="Permalink"></a></h3><pre><code class="language-julia">get_pedigree(founders)</code></pre><pre><code class="nohighlight">5×3 LinearAlgebra.Adjoint{Int64,Array{Int64,2}}:
1  0  0
2  0  0
3  0  0
4  0  0
5  0  0</code></pre><h3 id="Minor-Allele-Frequencies-(MAF)"><a class="docs-heading-anchor" href="#Minor-Allele-Frequencies-(MAF)">Minor Allele Frequencies (MAF)</a><a id="Minor-Allele-Frequencies-(MAF)-1"></a><a class="docs-heading-anchor-permalink" href="#Minor-Allele-Frequencies-(MAF)" title="Permalink"></a></h3><p>In the case where we have 3 QTLs out of 4 markers, we want to compare their allel frequencies.</p><pre><code class="language-julia">get_QTLs(founders) |&gt; get_MAF</code></pre><pre><code class="nohighlight">3-element Array{Float64,1}:
 0.5
 0.3
 0.5</code></pre><pre><code class="language-julia">get_MAF(founders)</code></pre><pre><code class="nohighlight">4-element Array{Float64,1}:
 0.5
 0.2
 0.3
 0.5</code></pre></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../build_phenome/">« Build Phenome</a><a class="docs-footer-nextpage" href="../mate/">Mate »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 0.27.3 on <span class="colophon-date" title="Wednesday 14 July 2021 11:54">Wednesday 14 July 2021</span>. Using Julia version 1.5.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
